vesc_hw_interface/src/vesc_wheel_controller.cpp

```diff
@@ -53,132 +68,94 @@ void VescWheelController::init(ros::NodeHandle nh, VescInterface* interface_ptr)
+  error_dt_ = target_velocity_ - current_vel;
+  error_ = (target_steps_ - position_steps_) * 2.0 * M_PI / (num_rotor_poles_ * num_hall_sensors_);
   error_integ_prev_ = error_integ_;
   error_integ_ += (error_ / control_rate_);
-  double duty = (kp_ * error_ + ki_ * error_integ_ + kd_ * error_dt_);
 
+  // limit integral
+  if (ki_ * error_integ_ > i_clamp_)
+  {
+    error_integ_ = i_clamp_ / ki_;
+  }
+  else if (ki_ * error_integ_ < -i_clamp_)
+  {
+    error_integ_ = -i_clamp_ / ki_;
+  }
+  const double u_p = kp_ * error_;
+  const double u_d = kd_ * error_dt_;
+  const double u_i = ki_ * error_integ_;
+  double u = u_p + u_d + u_i;
+
+  // limit duty value
   if (antiwindup_)
   {
-    if (duty > duty_limiter_)
+    if (u > duty_limiter_)
     {
-      duty = duty_limiter_;
-      if (error_integ_ > error_integ_prev_)
+      u = duty_limiter_;
+      if (error_integ_ > 0)
       {
-        error_integ_ = error_integ_prev_;
-        duty = (kp_ * error_ + ki_ * error_integ_ + kd_ * error_dt_);
+        error_integ_ = std::max(0.0, (duty_limiter_ - u_p - u_d) / ki_);
       }
     }
-    else if (duty < -duty_limiter_)
+    else if (u < -duty_limiter_)
     {
-      duty = -duty_limiter_;
-      if (error_integ_ < error_integ_prev_)
+      u = -duty_limiter_;
+      if (error_integ_ < 0)
       {
-        error_integ_ = error_integ_prev_;
-        duty = (kp_ * error_ + ki_ * error_integ_ + kd_ * error_dt_);
+        error_integ_ = std::min(0.0, (-duty_limiter_ - u_p - u_d) / ki_);
       }
     }
-    if (ki_ * error_integ_ > i_clamp_)
-    {
-      error_integ_ = i_clamp_ / ki_;
-    }
-    else if (ki_ * error_integ_ < -i_clamp_)
-    {
-      error_integ_ = -i_clamp_ / ki_;
-    }
-  }
-
-  // limit duty value
-  duty = std::clamp(duty, -duty_limiter_, duty_limiter_);
-  interface_ptr_->setDutyCycle(fabs(target_velocity) < 0.0001 ? 0.0 : duty);
-}
-
-double VescWheelController::counterTD(const double count_in, bool reset)
-{
-  if (reset)
-  {
-    counter_changed_single_ = 1;
-    for (int i = 0; i < 10; i++)
-    {
-      counter_changed_log_[i][0] = static_cast<uint16_t>(count_in);
-      counter_changed_log_[i][1] = 100;
-      counter_td_tmp_[i] = 0;
-    }
-    return 0.0;
-  }
-
-  if (counter_changed_log_[0][0] != static_cast<uint16_t>(count_in))
-  {
-    for (int i = 0; i < 10; i++)
-    {
-      counter_changed_log_[10 - i][0] = counter_changed_log_[9 - i][0];
-    }
-    counter_changed_log_[0][0] = static_cast<uint16_t>(count_in);
-    counter_changed_log_[0][1] = counter_changed_single_;
-    counter_changed_single_ = 1;
   }
   else
   {
-    if (counter_changed_single_ > counter_changed_log_[0][1])
-    {
-      counter_changed_log_[0][1] = counter_changed_single_;
-    }
-    if (counter_changed_single_ < 100)
-    {
-      counter_changed_single_++;
-    }
+    u = std::clamp(u, -duty_limiter_, duty_limiter_);
```

Comment:
Setting a duty ratio limit is essentially different from performing an I-control anti-windup.
This is because duty ratio limits are intended to limit motor power, while anti-windup is an algorithm to improve I-control response within a limited power output.
Therefore, even without anti-windup, it is necessary to limit the duty ratio.

Reply:
ok, make sense